### ISSUE TITLE
Test failure when cryptography built against OpenSSL 3.1 is run with 3.2

### ISSUE THREAD

**Issue by h-vetinari** [pyca/cryptography#9950]
Hey all

In conda-forge, we unbundle openssl and ship it separately from cryptography. To ensure we have the "right" OpenSSL available, we follow the upstream API/ABI-guarantees. For example, during the 1.1.1x series, building against 1.1.1h generated a runtime constraint of `openssl >=1.1.1h,<1.2`. With the changes to the versioning scheme in OpenSSL 3 (and their promise to stay ABI-compatible), we've loosened this a bit.

In particular, we pin only "openssl 3" at build time, which then generates `openssl >={actual_version_used_at_build_time},<4` for dependent packages.

As such, the recent release of OpenSSL 3.2 meant that packages started getting built against 3.2, and then requiring `>=3.2.0,<4` at runtime. So far, that is all in the regular course of things, only that OpenSSL 3.2 ended up [breaking](https://github.com/conda-forge/postgresql-feedstock/issues/178) postgres in various ways that are however not considered a regression by upstream (i.e. postgres was "misusing" functionality in a way that ended up breaking).

This now means we have to change the build-time constraint of cryptography (and possibly other packages) to explicitly build against "openssl 3.1", so that the runtime-requirement has a lower bound of 3.1.x, and thus stays compatible with the current postgres builds.

That normally shouldn't be a problem AFAIU, so I was surprised that building cryptography against openssl 3.1 and then running it against 3.2 has one test failure:
```
================= 1 failed, 3007 passed, 60 skipped in 46.45s ==================
```

<details>

More details in https://github.com/conda-forge/cryptography-feedstock/pull/129, resp. this [CI run](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=834933&view=results)

```
================================== FAILURES ===================================
_______________ TestRSADecryption.test_decrypt_invalid_decrypt ________________

self = <tests.hazmat.primitives.test_rsa.TestRSADecryption object at 0x00000212139b4bf0>
rsa_key_2048 = <cryptography.hazmat.backends.openssl.rsa._RSAPrivateKey object at 0x0000021219032f38>
backend = <OpenSSLBackend(version: OpenSSL 3.2.0 23 Nov 2023, FIPS: False, Legacy: True)>

    @pytest.mark.supported(
        only_if=lambda backend: (
            backend.rsa_encryption_supported(padding.PKCS1v15())
            and not backend._lib.Cryptography_HAS_IMPLICIT_RSA_REJECTION
        ),
        skip_message="Does not support PKCS1v1.5.",
    )
    def test_decrypt_invalid_decrypt(
        self, rsa_key_2048: rsa.RSAPrivateKey, backend
    ):
        private_key = rsa_key_2048
        with pytest.raises(ValueError):
>           private_key.decrypt(b"\x00" * 256, padding.PKCS1v15())
E           Failed: DID NOT RAISE <class 'ValueError'>

tests\hazmat\primitives\test_rsa.py:1794: Failed
```

</details>

Note that because the SAT solver will prefer the newest packages not precluded by other constraints, this is probably happening in many conda environments where someone had a cryptography build from before the release of OpenSSL 3.2, and then got just the OpenSSL 3.2 upgrade on the next environment manipulation[^1].

Summing up:
| OpenSSL at<br/>build time | OpenSSL at<br/>runtime | cryptography test<br/>suite passes |
|-|-|-|
| 3.1 | 3.1 | ✅ |
| 3.2 | 3.2 | ✅ |
| 3.1 | 3.2 | ❌ |

I'm not sure how this kind of cross dependence creeps in. Is it expected that the build-time version needs to match the runtime version, aside from basic things like availability of the correct symbols (i.e. what's covered by OpenSSL's ABI guarantees)?

[^1]: conda will [always](https://github.com/conda/conda/blob/23.11.0/conda/base/constants.py#L122-L126) try to upgrade openssl, but not necessarily cryptography

**Comment by alex** [comment]
Yes, you can skip it -- if the tests were compiled against 3.2 it would already be skipped :-)

**Comment by h-vetinari** [comment]
OK cool. I haven't fully groked why the tests need to be affected by a compile-time (rather than runtime) decision, but for now this should the problem. Thanks a lot for the help!
